### PR TITLE
Use the new I/O safety traits, using rustix

### PR DESCRIPTION
**Pull request by sunfishcode**
This implements the idea in https://github.com/yoshuawuyts/fd-lock/issues/12.

Switch public APIs from `AsRawFd`/`AsRawHandle` to the new I/O safety
traits `AsFd`/`AsHandle`. On Unix platforms, use `rustix` to avoid
manipulating raw file descriptors altogether.

This uses `AsFd` and `AsHandle` from the `io-lifetimes` crate, which are
identical to traits which will hopefully be landing in libstd soon.

### CODE REVIEW

**Comment by sunfishcode** [comment]
As a datapoint, [rustyline](https://github.com/kkawakam/rustyline) is a popular user of fd-lock, and it builds and runs all its tests successfully, using fd-lock from this branch and no other changes.

**Comment by sunfishcode** [comment]
As another datapoint, [cargo-spellcheck](https://github.com/drahnr/cargo-spellcheck/pull/200) is another popular user of fd-lock. Porting it to this branch just involved [updating it to fd-lock 3](https://github.com/drahnr/cargo-spellcheck/pull/200), and explicitly enabling the ["fs-err"](https://github.com/sunfishcode/io-lifetimes/blob/main/Cargo.toml#L35) feature in io-lifetimes, which adds [`AsFd`](https://docs.rs/io-lifetimes/0.2.4/io_lifetimes/trait.AsFd.html)/[`AsHandle`](https://docs.rs/io-lifetimes/0.2.4/x86_64-pc-windows-msvc/io_lifetimes/trait.AsHandle.html) implementations for [`fs_err::File`](https://docs.rs/fs-err/2.6.0/fs_err/struct.File.html), and it builds and runs all its tests successfully, with no other changes.

Once `AsFd`/`AsHandle` are stablized in libstd, it will make sense to submit `AsFd`/`AsHandle` implementations to upstream [fs-err](https://github.com/andrewhickman/fs-err) (as well as other popular crates), which will avoid the need to manually enable features for them.

**Review by yoshuawuyts — APPROVED** [review]
This looks amazing! Thank you so much!

**Comment by yoshuawuyts** [comment]
@sunfishcode I've invited you as an owner of both the git repo, and the crate on crates.io. Feel free to merge this and publish a new major if and when you think it's ready!

**Comment by sunfishcode** [comment]
My thought here was to wait until `AsFd` et al have landed in std and stabilized first.

If we land it as-is, some users will have to add an explicit dependency on io-lifetimes to enable the features they need, as the cargo-spellcheck example above does. We could avoid that if we submit `io_lifetimes::AsFd` impls upstream to `fs-err` and other major crates, but then we'd create more churn and compatibility problems when we eventually want to replace those impls with `std` impls once they're stable.

**Comment by sunfishcode** [comment]
I've now added a mode to io-lifetimes to use the new std types and traits on nightly. fd-lock's and rustyline's tests pass under this mode.

cargo-spellcheck needs `AsFd` implemented for `fs_err::File`, but io-lifetimes can't provide that in this mode due to the orphan rule, so I haven't tested it with the new mode yet.

**Comment by sunfishcode** [comment]
This is now rebased on #21, so now the diff just contains the `AsRawFd` to `AsFd` changes, making it clear what this change is all about :-).

**Comment by polarathene** [comment]
If I understand correctly, the [I/O safety traits are stabilized for a Rust 1.63](https://github.com/rust-lang/rust/pull/95118#event-6813166988) release? Will this become ready for merging and a release of `fd-lock` not long after as a result?

**Comment by sunfishcode** [comment]
When io_safety lands on stable, the next step I'm picturing is to start contributing `AsFd`/`From<OwnedFd>`/`From<T> for OwnedFd` impls to types in popular crates. For example, at least one of fd-lock's users calls fd-lock with an `fs_err::File`, so we'll need an `AsFd` impl for `fs_err::File`. The io-lifetimes crate has an impl for its own `AsFd` for `fs_err::File`, but to use the `AsFd` in std, we need the fs-err crate to have its own impl.

I've now posted a roadmap for how I envision this process working: https://github.com/sunfishcode/io-lifetimes/issues/38

**Comment by sunfishcode** [comment]
fs-err as of version 2.8.0 has `AsFd` impls, so this is now ready to go!

This changes APIs using `AsRawFd` to use `AsFd`, which isn't entirely semver-compatible, so this will require a semver major bump for fd-lock, to version 4.

**Comment by sunfishcode** [comment]
This is now released in fd-lock 4.0.0.

**Pull request merged**
